src/py/flwr/cli/build.py

```diff
@@ -105,38 +114,63 @@ def build(
-        for root, _, files in os.walk(app, topdown=True):
-            files = [
-                f
-                for f in files
-                if not ignore_spec.match_file(Path(root) / f)
-                and f != fab_filename
-                and Path(f).suffix in allowed_extensions
-                and f != "pyproject.toml"  # Exclude the original pyproject.toml
-            ]
-
-            for file in files:
-                file_path = Path(root) / file
-                archive_path = file_path.relative_to(app)
-                fab_file.write(file_path, archive_path)
-
-                # Calculate file info
-                sha256_hash = get_sha256_hash(file_path)
-                file_size_bits = os.path.getsize(file_path) * 8  # size in bits
-                list_file_content += f"{archive_path},{sha256_hash},{file_size_bits}\n"
-
-        # Add CONTENT and CONTENT.jwt to the zip file
-        fab_file.writestr(".info/CONTENT", list_file_content)
+    with tempfile.NamedTemporaryFile(suffix=".zip", delete=False) as temp_file:
+        temp_filename = temp_file.name
+
+        with zipfile.ZipFile(temp_filename, "w", zipfile.ZIP_DEFLATED) as fab_file:
+            write_to_zip(fab_file, "pyproject.toml", toml_contents)
+
+            # Continue with adding other files
+            for root, _, files in os.walk(app, topdown=True):
+                files = [
+                    f
+                    for f in files
+                    if not ignore_spec.match_file(Path(root) / f)
+                    and f != temp_filename
+                    and Path(f).suffix in allowed_extensions
+                    and f != "pyproject.toml"  # Exclude the original pyproject.toml
+                ]
+
+                for file in files:
+                    file_path = Path(root) / file
+
+                    # Read the file content manually
+                    with open(file_path, "rb") as f:
+                        file_contents = f.read()
+
+                    archive_path = file_path.relative_to(app)
+                    write_to_zip(fab_file, str(archive_path), file_contents)
+
+                    # Calculate file info
+                    sha256_hash = get_sha256_hash(file_path)
+                    file_size_bits = os.path.getsize(file_path) * 8  # size in bits
+                    list_file_content += (
+                        f"{archive_path},{sha256_hash},{file_size_bits}\n"
+                    )
+
+            # Add CONTENT and CONTENT.jwt to the zip file
+            write_to_zip(fab_file, ".info/CONTENT", list_file_content)
+
+    # Get hash of FAB file
+    content = Path(temp_filename).read_bytes()
+    fab_hash = hashlib.sha256(content).hexdigest()
+
+    # Set the name of the zip file
+    fab_filename = (
+        f"{conf['tool']['flwr']['app']['publisher']}"
+        f".{conf['project']['name']}"
+        f".{conf['project']['version'].replace('.', '-')}"
+        f".{fab_hash[:FAB_HASH_TRUNCATION]}.fab"
+    )
```

Comment:
I think we should create a util function for creating the fab file name instead of hand coded it here.